### ISSUE TITLE
feat(sampling): add top-n-sigma logit filtering (`top_n_sigma`) with a batch-safe row-filter hook

### ISSUE THREAD

**Issue by inureyes** [lablup/mlxcel#1375]
## Summary

mlxcel has no top-n-sigma sampler. Top-n-sigma keeps only the tokens whose logit lies within `n` standard deviations of the row maximum (`logit >= max - n * std`, statistics taken over the vocabulary row of the raw logits), masks the rest to `-inf`, and then samples as usual. Because it thresholds logits rather than probabilities it is invariant to temperature and to the per-row additive shift of a log-softmax, and it is the standard way to get a "stable at high temperature" sampler. This issue adds the filter, a `top_n_sigma` request field on every HTTP endpoint, a `--top-n-sigma` CLI flag, and the row-filter hook that #1377 and #1373 reuse, while keeping requests on the single-dispatch fused batch path.

## Current behavior

`SamplingConfig` carries temperature / top_k / top_p / min_p / penalties / DRY / XTC and nothing else (`src/lib/mlxcel-core/src/generate.rs:810-874`). The per-step pipeline is `preprocess_logits_for_sampling` (token bias, repetition, DRY, frequency/presence, XTC) followed by the C++ `ffi::fused_sample` chain (temperature, top_k, top_p, min_p, categorical):

```rust
// src/lib/mlxcel-core/src/sampling.rs:302-312
let last_logits = preprocess_logits_for_sampling(logits, config, token_history, state);
let token = ffi::fused_sample(&last_logits, config.temperature, config.top_k, config.top_p, config.min_p);
```

`grep -rn top_n_sigma src/` returns nothing. The nearest existing code is the XTC pre-step, `apply_xtc_step` at `src/lib/mlxcel-core/src/sampling.rs:1419-1430`, which is also a logit mask applied before the fused chain, but XTC draws from the RNG and therefore disqualifies the batched fused path (`config_supports_fused_batch`, `sampling.rs:628-630`). The batched fast path compares rows through `FusedSampleParams` (`sampling.rs:581-615`), which today holds exactly `temperature`, `top_k`, `top_p`, `min_p`.

## Expected behavior

New `SamplingConfig` field `top_n_sigma: f32`, default `0.0` (disabled). Valid range `>= 0.0`; `0.0` disables. A negative value is rejected at the request layer with HTTP 400 `invalid_request_error`, message `top_n_sigma must be >= 0.0`.

Filter definition, per row of a `[B, V]` logit tensor, computed in float32 regardless of the logit dtype (a float16 sum over a 150k-entry vocabulary overflows to `inf`, which would make the threshold `-inf` and silently disable the filter):

```
f        = logits.astype(f32)                       # [B, V]
finite   = isfinite(f)                              # entries already masked by token bias / penalties stay masked
n        = sum(finite, axis=-1, keepdims)           # finite count per row
mean     = sum(where(finite, f, 0), -1, keepdims) / n
d        = where(finite, f - mean, 0)
std      = sqrt(sum(d * d, -1, keepdims) / n)       # population std (ddof = 0)
top      = max(f, -1, keepdims)                     # -inf never wins while a finite entry exists
thresh   = top - top_n_sigma * std
keep     = f >= thresh                              # -inf >= thresh is false, so masked stays masked
out      = where(keep, logits, -inf)                # original dtype preserved
```

Worked example for the row `[0, 1, 2, 3, 4]`: mean 2, std `sqrt(2) = 1.414`; `n = 1` keeps `[3, 4]` (threshold 2.586), `n = 2` keeps `[2, 3, 4]` (threshold 1.172), `n = 0` keeps only the max, `n = 100` keeps everything. Rows are independent: `[[0,1,2,3,4],[4,3,2,1,0]]` with `n = 1` keeps indices `[3, 4]` in row 0 and `[0, 1]` in row 1.

Position in the chain. The filter runs after the history-based penalties and before XTC, i.e. on the penalized, untempered logits:

```
token bias -> repetition -> DRY -> frequency/presence -> [top_n_sigma -> p_less -> typical_p] -> XTC -> fused(temperature, top_k, top_p, min_p) -> draw
```

`top_n_sigma` is first of the three new row filters (the bracketed order is fixed by this issue; the other two land in their own issues). The greedy path is untouched: when `temperature == 0.0 || top_k == 1` the filter is skipped entirely, so greedy output stays byte-identical.

Fused batch path. The filter is row-wise, history-free and RNG-free, so it stays on the single-dispatch batched path: `config_supports_fused_batch` keeps returning `true`, `FusedSampleParams` gains a `top_n_sigma` field compared bitwise in `matches`, and `batched_fused_sample` applies the filter to the `[B, V]` batch before `ffi::fused_sample`. Two rows with different `top_n_sigma` values fall back to the per-row loop, exactly like two rows with different `top_p` do today.

Speculative decoding: `effective_token_distribution` and `sample_token_with_distribution` go through `preprocess_logits_for_sampling`, so the #902 proposal / target distributions pick the filter up automatically and stay consistent with the sampler.

## Implementation plan

1. `src/lib/mlxcel-core/src/generate.rs`: add `pub top_n_sigma: f32` to `SamplingConfig` with doc comment; set `0.0` in `Default` and in `greedy()`.

2. `src/lib/mlxcel-core/src/sampling.rs`:
   - Add `pub(crate) fn top_n_sigma_filter(logits: &MlxArray, n_sigma: f32) -> UniquePtr<MlxArray>` implementing the formula above with `ffi::astype(.., dtype::FLOAT32)`, `ffi::isfinite`, `ffi::where_cond`, `ffi::sum_axis`, `ffi::max_axis`, `ffi::sqrt`, `ffi::greater_equal`, `ffi::full_f32(&[1], f32::NEG_INFINITY, FLOAT32)`. Use `ops::divide_scalar` / `ffi::divide` for the `/ n` steps. Add a `// Used by:` comment naming the two call sites below.
   - Add a shared hook `fn apply_row_filters(logits: &MlxArray, params: &FusedSampleParams) -> UniquePtr<MlxArray>` that returns the input unchanged (no new graph nodes) when every filter is disabled or when `params.temperature == 0.0 || params.top_k == 1`, and otherwise applies the enabled filters in the fixed order `top_n_sigma`, `p_less`, `typical_p` (the latter two are added by their own issues; leave the match arms for them in place as comments so the order is pinned here).
   - Call the hook from `preprocess_logits_for_sampling` immediately before the XTC block (after the frequency/presence block), passing `FusedSampleParams::from_config(config)`.
   - `FusedSampleParams`: add `pub top_n_sigma: f32`; populate in `from_config`; compare with `to_bits()` in `matches`.
   - `batched_fused_sample`: after `ffi::slice_last_logits`, call `apply_row_filters(&last_logits, params)` and feed the result to `ffi::fused_sample`.
   - Leave `config_supports_fused_batch` unchanged (the filter is fused-eligible); extend its doc comment to say why.

3. Request plumbing (mirror the `xtc_probability` wiring exactly, request-only with no server default, absent resolves to `0.0`):
   - `src/server/types/request.rs` `SamplingParams`: `pub top_n_sigma: Option<f32>`.
   - `src/server/request_options.rs` `RequestOptionOverrides`: `pub top_n_sigma: Option<f32>`; in `build_server_generate_options` resolve `top_n_sigma: overrides.top_n_sigma.unwrap_or(0.0)` into `ResolvedSamplingParams`.
   - `src/execution/sampling.rs` `ResolvedSamplingParams` + `build_sampling_config`: thread the field through both the greedy and the stochastic branch.
   - `src/server/routes/chat.rs`: add `pub(crate) fn validate_top_n_sigma(value: Option<f32>) -> Result<(), &'static str>` next to `validate_xtc_params` (`chat.rs:1123`), rejecting `Some(v)` with `v < 0.0` or `!v.is_finite()`. Call it where `validate_xtc_params` is called in `chat.rs:259`, `completions.rs:169`, `responses.rs:109`, and map `params.top_n_sigma` in the `RequestOptionOverrides` literal at `chat.rs:1330`.
   - `src/server/routes/native_completion.rs:287`: map `request.top_n_sigma` (add `pub top_n_sigma: Option<f32>` to `NativeCompletionRequest`; the llama-server `/completion` schema carries this exact field name).
   - `src/distributed/disaggregated/serving_protocol.rs:278`: add the field to the wire struct so a disaggregated decode node honors it.
   - `src/bin/bench_decode.rs:295`: initialise to `0.0`.

4. CLI: `src/main.rs` `SamplingOptions` (line 612): `#[arg(long = "top-n-sigma", alias = "top-nsigma", default_value_t = 0.0, value_name = "FLOAT")] pub(crate) top_n_sigma: f32`, validated `>= 0.0` with a clap value parser. Thread through `resolved_cli_sampling_params` in `src/commands/generate.rs:910` and the chat equivalent in `src/commands/chat.rs:206`.

5. Docs: add the field to the sampling-parameters table in `docs/` where `xtc_probability` is documented, and to `mlxcel-server --help`.

## Validation

(a) Unit tests, appended to the existing `#[cfg(test)] mod tests` in `src/lib/mlxcel-core/src/sampling.rs` (the file's convention) and to `src/server/request_options_tests.rs` / `src/execution/sampling_tests.rs`:

- `top_n_sigma_filter_keeps_within_n_std`: row `[0,1,2,3,4]`, `n = 1.0` keeps exactly indices `{3, 4}`; `n = 2.0` keeps `{2, 3, 4}`; `n = 0.0` keeps `{4}`; `n = 100.0` keeps all.
- `top_n_sigma_filter_rows_independent`: two-row input above keeps `{3,4}` and `{0,1}`.
- `top_n_sigma_filter_ignores_neg_inf_entries`: row `[0, 1, 2, 3, 4, -inf]` produces the same kept set as without the `-inf` entry, and the `-inf` entry stays `-inf`.
- `top_n_sigma_filter_f16_large_vocab_does_not_overflow`: `[1, 152000]` float16 normal logits scaled by 3, `n = 1.0`: kept count is strictly between 0 and V.
- `top_n_sigma_skipped_on_greedy`: `temperature = 0.0` with `top_n_sigma = 1.0` returns the argmax of the unfiltered row and adds no graph nodes (compare `sample_token_optimized` output with the baseline config).
- `config_supports_fused_batch_true_for_top_n_sigma` and `fused_sample_params_matches_compares_top_n_sigma`.
- `batched_fused_sample_honors_top_n_sigma`: 64 identical rows `[0,1,2,3,4]`, `temperature 1.0`, `top_n_sigma 1.0`: every sampled id is in `{3, 4}`.
- `effective_token_distribution_zeroes_top_n_sigma_masked_tokens`: the #902 distribution has zero mass outside `{3, 4}`.
- Request layer: `chat_rejects_negative_top_n_sigma`, `completions_rejects_negative_top_n_sigma`, `responses_rejects_negative_top_n_sigma`, `native_completion_maps_top_n_sigma`, `build_sampling_config_threads_top_n_sigma_in_both_branches`.

(b) Real checkpoint: `mlx-community/Qwen3-4B-4bit`.

```
./target/release/mlxcel generate -m models/Qwen3-4B-4bit -p "Write a limerick about a lighthouse." -n 64 --temp 2.0 --top-n-sigma 1.0 --seed 1
./target/release/mlxcel generate -m models/Qwen3-4B-4bit -p "Write a limerick about a lighthouse." -n 64 --temp 2.0 --seed 1
```

Acceptance: the first run is fluent English at temperature 2.0 while the second degenerates (the documented property of the filter); greedy output (`--temp 0`) is token-identical with and without `--top-n-sigma 1.0`. Server: a `/v1/chat/completions` request with `"top_n_sigma": 1.0, "temperature": 2.0` returns fluent text and the batch scheduler log shows the fused batch dispatch (not the per-row fallback) for a batch of such requests; `"top_n_sigma": -1` returns 400.

## Acceptance criteria

- [x] `SamplingConfig::top_n_sigma` exists, defaults to `0.0`, and is honored by `sample_token_optimized`, `batched_fused_sample`, `effective_token_distribution`, and `sample_token_with_distribution`
- [x] Disabled value leaves every existing token stream byte-identical (no new graph nodes when `top_n_sigma == 0.0`)
- [x] Field accepted on `/v1/chat/completions`, `/v1/completions`, `/v1/responses`, and the disaggregated wire protocol; negative values rejected with 400. The native `/completion` spelling is rescoped to #1436: the llama-compat gate ties `NativeCompletionRequest` field declarations to the `field:top_n_sigma` manifest entry, which #1436 owns under the epic #1431 shard-ownership rule, so the field and its manifest flip land together there
- [x] `--top-n-sigma` on `mlxcel generate` and `mlxcel chat`
- [x] Uniform `top_n_sigma` batches stay on the fused batch path; mixed values fall back per row
- [x] Unit tests listed above pass; real-checkpoint check above passes
- [x] cargo test --workspace --profile test-fast --features metal,accelerate passes
- [x] cargo clippy --workspace --all-targets -- -D warnings and cargo fmt --all -- --check pass

## Out of scope

- A server-wide default flag (`--top-nsigma` on `mlxcel-server`); the field is request-only like XTC.
- Moving temperature after the filters (tracked by #1379); top-n-sigma is invariant to that change.
